src/box/box.cc

```diff
@@ -1805,6 +1805,218 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+		.deadline = deadline,
+		.is_timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.syncs[replica->id] != 0)
+			continue;
+		if (relay_next_sync(replica->relay, &data.syncs[replica->id],
+				    deadline) != 0) {
+			/* Timed out. */
+			trigger_clear(&on_ack);
+			trigger_clear(&on_relay_thread_start);
+			return -1;
+		}
+	}
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon)
+			continue;
+		uint64_t sync = data.syncs[replica->id];
+		struct relay *relay = replica->relay;
+
+		/* Might be already accounted by on_ack trigger. */
+		if (!bit_test(&data.collected_vclock_map, replica->id) &&
+		    sync > 0 && relay_sync(relay) >= sync) {
+			vclock_max_ignore0(confirmed_vclock,
+					   relay_vclock(relay));
+			bit_set(&data.collected_vclock_map, replica->id);
+		}
+	}
+
+	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
+	       !data.is_timed_out) {
+		if (fiber_yield_deadline(deadline))
+			break;
+	}
+
+	trigger_clear(&on_ack);
+	trigger_clear(&on_relay_thread_start);
+	if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
+		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
+			 "timed out waiting for remote node vclocks");
+		return -1;
+	}
+	return 0;
+}
+
+/** box_wait_vclock trigger data. */
+struct box_wait_vclock_data {
+	/** Whether the request is finished. */
+	bool is_ready;
+	/** The vclock to wait for. */
+	const struct vclock *vclock;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+};
+
+static int
+box_wait_vclock_f(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct box_wait_vclock_data *data =
+		(struct box_wait_vclock_data *)trigger->data;
+	if (vclock_compare_ignore0(data->vclock, &replicaset.vclock) <= 0) {
+		data->is_ready = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/**
+ * Wait until this instance's vclock reaches @a vclock or @a deadline is
+ * reached.
+ */
+static int
+box_wait_vclock(const struct vclock *vclock, double deadline)
+{
+	if (vclock_compare_ignore0(vclock, &replicaset.vclock) <= 0)
+		return 0;
+	struct trigger on_wal_write;
+	struct box_wait_vclock_data data = {
+		.is_ready = false,
+		.vclock = vclock,
+		.waiter = fiber(),
+	};
+	trigger_create(&on_wal_write, box_wait_vclock_f, &data, NULL);
+	trigger_add(&wal_on_batch_write, &on_wal_write);
+	do {
+		if (fiber_yield_deadline(deadline))
+			break;
+	} while (!data.is_ready);
```

Comment:
Should we also break if the fiber is cancelled? Same for the while-loop in `box_collect_confirmed_vclock()`.

Reply:
Yes, probably. Altough `relay_next_sync` can't be interrupted anyway. Because `cbus_call_timeout` can't.
But you're right. We should at least let the user cancel the fiber while it waits for already known syncs or vclocks.

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 8a793df7d..6a1485a5d 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1933,13 +1934,18 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
 	}
 
 	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
-	       !data.is_timed_out) {
+	       !data.is_timed_out && !fiber_is_cancelled()) {
 		if (fiber_yield_deadline(deadline))
 			break;
 	}
 
 	trigger_clear(&on_ack);
 	trigger_clear(&on_relay_thread_start);
+	if (fiber_is_cancelled()) {
+		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
+			 "fiber is cancelled");
+		return -1;
+	}
 	if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
 		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
 			 "timed out waiting for remote node vclocks");
@@ -1991,10 +1997,16 @@ box_wait_vclock(const struct vclock *vclock, double deadline)
 	do {
 		if (fiber_yield_deadline(deadline))
 			break;
-	} while (!data.is_ready);
+	} while (!data.is_ready && !fiber_is_cancelled());
 	trigger_clear(&on_wal_write);
+	if (fiber_is_cancelled()) {
+		diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
+			 "fiber is cancelled");
+		return -1;
+	}
 	if (!data.is_ready) {
-		diag_set(ClientError, ER_TIMEOUT);
+		diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
+			 "timed out");
 		return -1;
 	}
 	return 0;
diff --git a/src/box/errcode.h b/src/box/errcode.h
index 733d3f137..4f5f7ef73 100644
--- a/src/box/errcode.h
+++ b/src/box/errcode.h
@@ -297,6 +297,7 @@ struct errcode_record {
 	/*242 */_(ER_NO_ELECTION_QUORUM,	"Not enough peers connected to start elections: %d out of minimal required %d")\
 	/*243 */_(ER_SSL,			"%s") \
 	/*244 */_(ER_SPLIT_BRAIN,		"Split-Brain discovered: %s") \
+	/*245 */_(ER_VCLOCK_WAIT,		"Failed to wait for vclock %s: %s")\
 
 /*
  * !IMPORTANT! Please follow instructions at start of the file
```
</details>